File: examples/extensions/template.js
#!/usr/bin/env node
// llm extension template (js, pi-compatible) — copy to
// ~/.llm/extensions/<name> and edit the USER SECTION at the bottom.
// The shim below speaks the agent's stdio protocol; the user section is
// written in pi's extension API style (registerTool / registerCommand /
// on(event)), which maps onto it directly.

// ---------------------------------------------------------------- protocol --
const TOOLS = {};    // name -> {description, parameters, execute}
const COMMANDS = {}; // name -> handler(argText) -> string
const EVENTS = {};   // event name -> [handler(params) -> reply-or-undefined]

function reply(obj) {
  process.stdout.write(JSON.stringify(obj) + "\n");
}

// ---------------------------------------------------------------- pi shim ---
// The subset of pi's extension API that can be honored out-of-process:
// tools, slash commands, and event hooks. UI / editor / hotkey APIs are
// not available and calling them raises with a clear message.
const pi = {
  registerTool(def) {
    if (!def || !def.name) throw new Error("registerTool: name required");
    TOOLS[def.name] = {
      description: def.description || "",
      parameters: def.parameters || { type: "object", properties: {} },
      execute: def.execute || (() => ""),
    };
  },

  registerCommand(name, def) {
    const handler = typeof def === "function" ? def : def && def.handler;
    if (!handler) throw new Error("registerCommand: handler required");
    COMMANDS[name] = handler;
  },

  on(event, handler) {
    (EVENTS[event] = EVENTS[event] || []).push(handler);
  },

  unsupported(api) {
    throw new Error(
      `this pi extension uses "${api}", which is not available to ` +
        `out-of-process extensions (tools, commands and event hooks are)`
    );
  },

  run() {
    let buffer = "";
    process.stdin.setEncoding("utf8");
    process.stdin.on("data", (chunk) => {
      buffer += chunk;
      let idx;
      while ((idx = buffer.indexOf("\n")) >= 0) {
        const line = buffer.slice(0, idx).trim();
        buffer = buffer.slice(idx + 1);
        if (line) dispatch(line);
      }
    });
  },
};

async function dispatch(line) {
  let req;
  try {
    req = JSON.parse(line);
  } catch {
    return;
  }
  const kind = req.type;
  if (kind === "initialize") {
    reply({
      id: req.id,
      result: {
        tools: Object.entries(TOOLS).map(([name, t]) => ({
          name,
          description: t.description,
          parameters: t.parameters,
        })),
        commands: Object.keys(COMMANDS),
        events: Object.keys(EVENTS),
      },
    });
  } else if (kind === "call_tool") {
    const tool = TOOLS[req.name];
    try {
      if (!tool) throw new Error(`unknown tool ${req.name}`);
      const out = await tool.execute(req.args || {});
      reply({ id: req.id, result: typeof out === "string" ? out : JSON.stringify(out) });
    } catch (e) {
      reply({ id: req.id, result: `error: ${e.message}` });
    }
  } else if (kind === "run_command") {
    try {
      const handler = COMMANDS[req.name];
      if (!handler) throw new Error(`unknown command ${req.name}`);
      const out = await handler(req.args || "");
      reply({ id: req.id, result: typeof out === "string" ? out : JSON.stringify(out) });
    } catch (e) {
      reply({ id: req.id, result: `error: ${e.message}` });
    }
  } else if (kind === "event") {
    let result;
    for (const handler of EVENTS[req.name] || []) {
      try {
        const r = await handler(req.params || {});
        if (result === undefined && r && typeof r === "object") result = r;
      } catch (e) {
        process.stderr.write(`${req.name} handler failed: ${e.message}\n`);
      }
    }
    reply({ id: req.id, result: result ?? null });
  } else if (kind === "shutdown") {
    // let in-flight async handlers land their replies first
    setTimeout(() => process.exit(0), 100);
  } else {
    pi.unsupported(kind);
  }
}

// ---------------------------------------------------------- USER SECTION ----
// pi-style extension code goes here. Examples:

pi.registerTool({
  name: "now",
  description: "Current local time",
  parameters: { type: "object", properties: {} },
  execute: async () => new Date().toString(),
});

pi.on("tool_call", (params) => {
  // return { decision: "deny", reason: "no" } to block a call, or
  // { args: {...} } to rewrite its arguments
  return undefined;
});

pi.run();

File: examples/extensions/template.py
#!/usr/bin/env python3
"""llm extension template (python) — copy to ~/.llm/extensions/<name> and
edit the USER SECTION below. The protocol loop above talks to the agent
host; nothing below needs to know about it."""

import json
import os
import sys

# ---------------------------------------------------------------- protocol --
TOOLS = {}       # name -> (description, schema, handler(args_dict) -> str)
COMMANDS = {}    # name -> handler(arg_text) -> str
EVENTS = {}      # event name -> [handler(params_dict) -> reply-or-None]


def reply(obj):
    sys.stdout.write(json.dumps(obj) + "\n")
    sys.stdout.flush()


def tool(name, description, parameters=None):
    """Register a tool; the handler receives the arguments dict and
    returns the result text."""
    def wrap(handler):
        TOOLS[name] = (
            description,
            parameters or {"type": "object", "properties": {}},
            handler,
        )
        return handler

    return wrap


def on(event):
    """Subscribe to an agent event. A tool_call handler may return
    {"decision": "deny", "reason": ..} or {"args": {...}} to rewrite."""
    def wrap(handler):
        EVENTS.setdefault(event, []).append(handler)
        return handler

    return wrap


def command(name):
    """Register a /slash command; the handler receives the argument text
    and returns the reply text."""
    def wrap(handler):
        COMMANDS[name] = handler
        return handler

    return wrap


def run():
    while True:
        line = sys.stdin.readline()
        if not line:
            break
        line = line.strip()
        if not line:
            continue
        try:
            req = json.loads(line)
        except ValueError:
            continue
        kind = req.get("type")
        if kind == "initialize":
            reply({
                "id": req["id"],
                "result": {
                    "tools": [
                        {"name": n, "description": d, "parameters": s}
                        for n, (d, s, _) in TOOLS.items()
                    ],
                    "commands": list(COMMANDS),
                    "events": list(EVENTS),
                },
            })
        elif kind == "call_tool":
            try:
                handler = TOOLS[req["name"]][2]
                out = handler(req.get("args") or {})
                reply({"id": req["id"], "result": str(out)})
            except Exception as e:  # tool errors are results, not crashes
                reply({"id": req["id"], "result": f"error: {e}"})
        elif kind == "run_command":
            try:
                out = COMMANDS[req["name"]](req.get("args") or "")
                reply({"id": req["id"], "result": str(out)})
            except Exception as e:
                reply({"id": req["id"], "result": f"error: {e}"})
        elif kind == "event":
            result = None
            for handler in EVENTS.get(req["name"], []):
                try:
                    r = handler(req.get("params") or {})
                    if result is None and isinstance(r, dict):
                        result = r
                except Exception as e:
                    sys.stderr.write(f"{req['name']} handler failed: {e}\n")
            reply({"id": req["id"], "result": result})
        elif kind == "shutdown":
            break


# ----------------------------- USER SECTION ---------------------------------
# Register tools, commands and event hooks here. Examples:

@tool("now", "Current local time")
def now(args):
    import time
    return time.strftime("%Y-%m-%d %H:%M:%S")


@on("tool_call")
def gate(params):
    # return {"decision": "deny", "reason": "no"} to block a call, or
    # {"args": {...}} to rewrite its arguments
    return None


run()
